File: 课程网站/microlectures/2.3-frequency-response/source/first_experiment.m
% Standalone MATLAB script. Change w to compare frequencies.
n=0:79; w=0.2*pi; h=ones(1,5)/5;
x=cos(w*n); y=filter(h,1,x);
H=sum(h.*exp(-1i*w*(0:4))); steady=real(H*exp(1i*w*n));
c=[cos(w*n(5:end)).',sin(w*n(5:end)).']\y(5:end).';
err=max(abs(y(5:end)-steady(5:end))); assert(err<1e-12);
fprintf('Gain %.9f, fitted %.9f, steady error %.3g\n',abs(H),norm(c),err);
if abs(H)>1e-10, fprintf('Phase %.9f rad\n',angle(H)); else, fprintf('Phase undefined at zero response\n'); end
f=figure('Visible','off'); plot(n,x,'.-',n,y,'o-',n,steady,'--');
xlim([0 30]); xlabel('n (sample index)'); ylabel('Amplitude');
legend('input','zero-history output','steady prediction'); grid on;
exportgraphics(f,fullfile(fileparts(mfilename('fullpath')),'first_experiment_matlab.png')); close(f);

File: 课程网站/microlectures/2.3-frequency-response/source/first_experiment.py
"""Standalone NumPy/Matplotlib experiment; outputs beside this script."""
from pathlib import Path
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
n=np.arange(80);w=.2*np.pi;h=np.ones(5)/5
x=np.cos(w*n);y=np.convolve(x,h)[:len(n)]
H=np.sum(h*np.exp(-1j*w*np.arange(5)));steady=np.real(H*np.exp(1j*w*n))
coef=np.linalg.lstsq(np.column_stack([np.cos(w*n[4:]),np.sin(w*n[4:])]),y[4:],rcond=None)[0]
error=float(np.max(np.abs(y[4:]-steady[4:])))
assert error<1e-12
print('gain=',abs(H),' fitted=',np.linalg.norm(coef),' phase=',np.angle(H) if abs(H)>1e-10 else 'undefined',' steady error=',error)
fig,ax=plt.subplots(figsize=(10,4));ax.plot(n,x,'.-',label='input');ax.plot(n,y,'o-',label='zero-history output');ax.plot(n,steady,'--',label='steady prediction');ax.axvspan(0,3.5,color='gray',alpha=.15);ax.set(xlim=(0,30),xlabel='n (sample index)',ylabel='Amplitude');ax.legend();fig.tight_layout();fig.savefig(Path(__file__).with_suffix('.png'));plt.close(fig)
